modules/correlation/radix.c

```diff
@@ -100,11 +100,14 @@ r_parser_nlstring(gchar *str, gint *len, const gchar *param, gpointer state, RPa
       /* drop CR before to LF */
       if (end - str >= 1 && *(end - 1) == '\r')
         end--;
-      *len = (end - str);
-      return TRUE;
     }
   else
-    return FALSE;
+    {
+      end = strchr(str, '\0');
```

Comment:
this would be equivalent to an strlen()

Reply:
This does not consume the NL character at the moment, so the pattern will need to contain that. This was originally meant to match strings in a formatted Windows event log, like this:

```
Subject:
 Security ID:  @NLSTRING:@
 Account Name:  @NLSTRING:@
 Account Domain:  @NLSTRING:.win.SubjectDomainName@
 Logon ID:  @NLSTRING:@
```

e.g. the newline was consumed literally by the pattern. I think this is still a good choice, so no need to consume the NL character (which is in contrast with what QSTRING does). 

I am thinking about the CR character though... We should have skipped that (e.g. not include it in the extracted value but skip it when processing the rest of the pattern). But this would be an incompatible change now, so probably not a good idea.

Reply:
One more thought: if `@NLSTRING@` does not consume the NL character itself and requires the pattern to subsequently match it, then the final NL character is a bit of an issue here as you NEED to expect an NL character to follow the `@NLSTRING@` in usual cases.

But this is a conflict:

1) normally `@NLSTRING@` needs to be followed by NL to continue matching
2) but at the end of the message, there's no subsequent NL character, so if you knew this, you could have written `@ANYSTRING@` instead.

So I am not sure this is a good idea after all. When we created NLSTRING, we should have consumed the NL (& CR for that matter) characters and then allow breaking the lines in the pattern string itself to make it readable, but  without matching the NL character literally. I am not sure if we can easily introduce such a syntax now.

The usual suspect to indicate continuation lines in various formats is the use of the backslash, but that backslash is now processed literally. The only 'escape' character is the '@' itself. We could do something like:

```
Subject:
 Security ID:  @NLSTRING2:@@\
 Account Name:  @NLSTRING2:@@\
 Account Domain:  @NLSTRING2:.win.SubjectDomainName@@\
 Logon ID:  @NLSTRING2:@@\

```

Notice the `@NLSTRING2@` parser, which would consume the NL characters and then the `@\` sequence to escape the following NL indicating it's just a line break in the pattern but not in the actual text. 

This is pretty ugly. But would be compatible.

Reply:
@NLSTRING@ does not consume the NL at the moment. It matches the NL and then does not include it in the match itself. It is matched as a part of the pattern.

If you look at `scl/windowseventlog/windowseventlog.xml`, it has this:

```
Subject:
 Security ID:  @NLSTRING:.win.TargetUserSid@
 Account Name:  @NLSTRING:.win.TargetUserName@
 Account Domain:  @NLSTRING:.win.SubjectDomainName@
 Logon ID:  @NLSTRING:.win.TargetLogonId@

```

NOTE the NL in the pattern _after the `@NLSTRING@` block. That is needed because the NLSTRING does not consume the NL character, rather it leaves it up to the pattern to parse. Therefore the pattern itself contains a - literal - NL character.

If `@NLSTRING@` was consuming it, then you wouldn't need the NL in the pattern, which I think would make it less readable.

Reply:
I am not sure to understand if there is an issue or not? :thinking: 

The behavior of this PR looks good AFAIAC (bellow I am useng `&#10;` and `&#13;` to replace CR and LF because CR are not displayed by GitHub).  Not consuming the LF (`\n`) allows to reproduce the structure of the message on multiple line as visible in your examples above (and mine before I wanted to show that my examples where mixing LF and CR+LF below),  and the normalization makes it more straightforward to write rules… at least for users of UNIX like operating systems :upside_down_face:.

Considering the following example, two test describe what is currently supported and works, while the two other do not pass with the current code of syslog-ng.

With this patch, all 4 tests pass.

```xml
<?xml version='1.0' encoding='UTF-8'?>
<patterndb version='4' pub_date='2017-08-23'>
    <ruleset name='Test' id='test'>
        <pattern>test</pattern>
        <rules>
            <rule id='test1'>
                <patterns>
                    <pattern>@NLSTRING:m1@&#10;@NLSTRING:m2@&#10;@NLSTRING:m3@&#10;</pattern>
                </patterns>
                <tags></tags>
                <values></values>
                <examples>
                    <example>
                        <test_message program='test'>foo&#10;bar&#13;&#10;baz&#10;</test_message>
                        <test_values>
                            <test_value name='m1'>foo</test_value>
                            <test_value name='m2'>bar</test_value>
                            <test_value name='m3'>baz</test_value>
                        </test_values>
                    </example>
                    <example>
                        <test_message program='test'>foo&#10;bar&#13;&#10;baz</test_message>
                        <!-- fail with current pdbtool test, no trailing \n -->
                        <test_values>
                            <test_value name='m1'>foo</test_value>
                            <test_value name='m2'>bar</test_value>
                            <test_value name='m3'>baz</test_value>
                        </test_values>
                    </example>
                </examples>
            </rule>
            <rule id='test2'>
                <patterns>
                    <pattern>@NLSTRING:m1@&#10;@NLSTRING:m2@&#10;@NLSTRING:m3@</pattern>
                </patterns>
                <tags></tags>
                <values></values>
                <examples>
                    <example>
                        <test_message program='test'>foo&#10;bar&#13;&#10;baz&#10;</test_message>
                        <test_values>
                            <test_value name='m1'>foo</test_value>
                            <test_value name='m2'>bar</test_value>
                            <test_value name='m3'>baz</test_value>
                        </test_values>
                    </example>
                    <example>
                        <test_message program='test'>foo&#10;bar&#13;&#10;baz</test_message>
                        <!-- fail with current pdbtool test, no trailing \n -->
                        <test_values>
                            <test_value name='m1'>foo</test_value>
                            <test_value name='m2'>bar</test_value>
                            <test_value name='m3'>baz</test_value>
                        </test_values>
                    </example>
                </examples>
            </rule>
        </rules>
    </ruleset>
</patterndb>
```

My guess is that this allow to remove the extra `\n` at the end of patterns and examples which slightly improve readability.

My use case is processing messages that are usualy a single info line, but sometimes an error followed by a backtrace I don't really need to match, so:

```xml
<?xml version='1.0' encoding='UTF-8'?>
<patterndb version='4' pub_date='2017-08-23'>
    <ruleset name='Test' id='test'>
        <pattern>test</pattern>
        <rules>
            <rule id='test'>
                <patterns>
                    <pattern>@NLSTRING:message@</pattern>
                </patterns>
                <tags></tags>
                <values></values>
                <examples>
                    <example>
                        <test_message program='test'>Everything is fine!</test_message>
                        <test_values>
                            <test_value name='message'>Everything is fine!</test_value>
                        </test_values>
                    </example>
                    <example>
                        <test_message program='test'>Critical failure, everything is NOT fine!
More info: https://www.youtube.com/watch?v=C7sE0agEVNg</test_message>
                        <test_values>
                            <test_value name='message'>Critical failure, everything is NOT fine!</test_value>
                        </test_values>
                    </example>
                </examples>
            </rule>
        </rules>
    </ruleset>
</patterndb>
```

Reply:
I am convinced.